### ISSUE TITLE
Progress Bar for Tray in Mac Menu bar

### ISSUE THREAD

**Issue by deadcoder0904** [electron/electron#13500]
## **Is your feature request related to a problem? Please describe.**
Yes

## **Describe the solution you'd like**
I want to show a progress bar in the menu bar like Battery App. Or something moving like https://www.progressbarosx.com

Idk if its already possible to do this using Electron as Google couldn't help :)

**Comment by adlk** [comment]
This is already possible by e.g. dynamically creating the progress bar as an SVG and export it as an datURI. From there, create an NativeImage (`nativeImage.createFromDataURL(dataURL)`) and use it for the trayIcon. 

With this approach you are not just adding support for a progress bar, instead you can always do whatever you want.

**Comment by deadcoder0904** [comment]
Wow didn't knew this was possible. Do you have any examples of some Electron app implementing this?

**Comment by deadcoder0904** [comment]
Cool I'll try implement a simple example with it & if it works, I'll close this issue :)

**Comment by deadcoder0904** [comment]
Hey @adlk, I'm trying to make a Proof Of Concept for it at https://github.com/deadcoder0904/end-of-life by making a Progress Bar going from 100% to 0% but as much I have researched SVG is not supported in Electron.

You can checkout [Tray docs](https://electronjs.org/docs/api/tray) which points to [nativeImage docs](https://electronjs.org/docs/api/native-image) which says only PNG & JPEG is supported. Have you ever used SVG in Electron App Icons since I found issues saying that its not supported?

**Comment by adlk** [comment]
This is the reason why you have to convert it to a dataURI first (of course a cached PNG would be fine as well).

**Comment by deadcoder0904** [comment]
Yeah its not happening using `nativeImage.createFromDataURL`

Checkout https://github.com/deadcoder0904/end-of-life/blob/master/index.js#L22. It only shows Tray Icon when submitted directly in the constructor of Tray like `new Tray('path/to/icon.png')`. As you know, it does not work directly for `.svg`.

Also, `nativeImage.createFromDataURL` does not work for either `.png` or `.svg`. Any suggestions or if you can clone the repo if you have time I'd appreciate it?

**Comment by adlk** [comment]
Well, for `createFromDataURL` you need to provide an actual dataURI instead of a path. If you want to test it with a path to your PNG you need to use `createFromPath()`.
Try this to get a datauri from your SVG: https://www.npmjs.com/package/datauri#synchronous-function 

This might be an interesting read for you: http://blog.teamtreehouse.com/using-data-uris-speed-website

**Comment by deadcoder0904** [comment]
I tried everything you said for the past 2 hours & it still is not working. I tried directly putting SVG into `datauri` package but it still doesn't work. `datauri` package wasn't working at first but I found a workaround by installing `shebang-loader` so it worked. Similar packages were not working. I tried 4-5 different packages.

```bash
App threw an error during load
Error: ENOENT: no such file or directory, open '/Desktop/end-of-life/dist/data/mime.types'
    at Object.fs.openSync (fs.js:577:3)
    at Object.module.(anonymous function) [as openSync] (ELECTRON_ASAR.js:172:20)
    at Object.fs.readFileSync (fs.js:483:33)
    at Object.fs.readFileSync (ELECTRON_ASAR.js:510:29)
    at module.exports (webpack:///./node_modules/.registry.npmjs.org/mimer/0.3.2/node_modules/mimer/lib/data/parser.js?:9:8)
    at Object.eval (webpack:///./node_modules/.registry.npmjs.org/mimer/0.3.2/node_modules/mimer/lib/exec.js?:62:193)
    at eval (webpack:///./node_modules/.registry.npmjs.org/mimer/0.3.2/node_modules/mimer/lib/exec.js?:17:37)
    at eval (webpack:///./node_modules/.registry.npmjs.org/mimer/0.3.2/node_modules/mimer/lib/exec.js?:20:2)
    at Object../node_modules/.registry.npmjs.org/mimer/0.3.2/node_modules/mimer/lib/exec.js (/Desktop/end-of-life/dist/main.js:121:76)
    at r (/Desktop/end-of-life/dist/main.js:1:265)
```

Here's the code https://github.com/deadcoder0904/end-of-life/blob/master/index.js#L26

**Comment by adlk** [comment]
The issue you've posted here is actually an issue with `mimer` and webpack. Find the mimer package in your node_modules and copy the `data` folder to `./dist`. This will solve your first problem. (You need to copy your assets to dist as well)

So let's assume you are dynamically generating an SVG. You first need to export it as a PNG and then either store it on disk or use `DataURI()`. If you look at the source below you'll find a perfectly working example.
<img width="205" alt="screen shot 2018-07-04 at 12 45 49" src="https://user-images.githubusercontent.com/3265004/42273025-37508a2e-7f88-11e8-91ba-4d9d4b062f6b.png">


```js
import path from "path";
import url from "url";
import { app, Tray, nativeImage, Menu } from "electron";
import { sync as DataURI } from "datauri";

import MainWindow from "./app/main_window";

// import icon from "./src/assets/progress-bar.svg";
// import icon from "./src/assets/btc.png";

let win = null;
let tray = null;

app.on("ready", () => {
  win = new MainWindow(
    url.format({
      pathname: path.join(__dirname, "./index.html"),
      protocol: "file:",
      slashes: true
    }),
    {
      show: true
    }
  );

  let trayIcon = nativeImage.createFromDataURL(
    DataURI(path.join(__dirname, "src/assets/btc.png"))
  );
  trayIcon = trayIcon.resize({ width: 16, height: 16 });
  tray = new Tray(trayIcon);
  tray.setTitle('<- look at me')

  win.on("closed", () => {
    win = null;
  });
});

// app.dock.hide();
```

**Bonus note:** your setup has a few quirks. Just use electron-webpack that solves pretty much everything for you.

**Comment by deadcoder0904** [comment]
Hey man, thanks for all the help. I changed my setup to `electron-webpack` which is not as much pain as my custom setup but I'll switch to `electron-forge` probably since its much simpler :)

I implemented exactly what you have shown & deleted the old repo & created a new one with same name 

🔗 https://github.com/deadcoder0904/end-of-life

But still only PNG works out & SVG does not work at all. I have setup everything in place so you can see for yourself by cloning it, its not working with SVG. No icon shows up with SVG but with PNG the BTC icon shows up. Any suggestions now❓

**Comment by adlk** [comment]
electron-forge is great too!

> do I still have to convert SVG to PNG to make it work

yes.

> Another thing is wouldn't it be inefficient to store a new file on disk for every percentage assuming this app is like Battery increasing 1% at a time or suppose if its Pomodoro Clock then 1 file per second sounds too much❓

**A:** If you work with a dynamically generated SVG, you don't necessarily have to store it on disk as you can export it from canvas directly to PNG. As it's basically in a server-side-rendered environment [jsdom](https://github.com/jsdom/jsdom) might be helpful.
```
var canvas = document.getElementById("mycanvas");
var img    = canvas.toDataURL("image/png");
```
**B:** If you just want to display a progress bar with a value from 0-100 you can pre-render them and cache them on disk. Even before building the app, there is no need to do this during runtime.  

**C:** You can merge images like combining a pre-rendered progress bar with a dynamic text that is created via a SVG.

The thing is, you can do whatever you want. There is just not a straight path.

**Comment by deadcoder0904** [comment]
I tried for the past 3 hours using Canvas but it gave me tons of errors. JSDOM installs well & works but canvas & node-canvas does not install at all & gives me some weird error. 

I tried implementing https://stackoverflow.com/questions/27230293/how-to-convert-svg-to-png-using-html5-canvas-javascript-jquery-and-save-on-serve using JSDOM but it doesn't print again no errors but doesn't print anything.

I think I am done with this now. If anyone else wants to tackle this issue, feel free to checkout my repo https://github.com/deadcoder0904/end-of-life

Thanks for all the help @adlk. Also, love Franz :)
